File: Sample/Source/Item.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
from typing import Optional, Union, List, Tuple
from Engine import Pair, Texture, IntRect
from Engine.Gameplay.Actors import Actor
from Global import GameMap, Manager
from . import Data, System
from .Infos.ItemInfo import ItemInfo
from Source.NodeFunctions.Player import MeetPlayer


@Meta(GeneralDataVars=[("ID", "Item")])
class Item(Actor, ItemInfo):
    r"""
    \brief Scene item entity.

    Bridges Actor (rendering/collision/movement) and ItemInfo (item data + event logic)
    via multiple inheritance.
    """

    ID: str = "FILL_IT_BY_YOURSELF"

    def __init__(
        self,
        texture: Optional[Union[Texture, List[Texture]]] = None,
        rect: Union[IntRect, Tuple[Pair[int], Pair[int]]] = None,
        tag: Optional[str] = None,
    ) -> None:
        r"""\brief Construct an item with actor rendering and item info.

        - \param texture Optional texture or list of textures for the actor sprite.
        - \param rect Optional texture rectangle or pair of position/size pairs.
        - \param tag Optional actor tag.
        """
        Actor.__init__(self, texture, rect, tag)
        self.initInfo(Data)

    def onCollision(self, other: List[Actor]) -> None:
        from Source.Scenes import Map

        player = MeetPlayer(other)
        if player:
            scene = Cast(Map, Cast(GameMap, self.getMap()).getScene())
            inst = scene.inst
            Manager.playSE(System.getGetSE())
            player.addItem(self.ID)
            if not inst.getCachedNewItem(self.ID):
                inst.setCachedNewItem(self.ID)
                scene.showMessage("", LOC("ITEM_NEW").format(name=self.name, desc=self.desc).replace("\\n", "\n"), "")
        super().onCollision(other)
        scene.recordDestroyedActor(self)
        self.destroy()

File: Sample/Scripts/Source/Item.lua
local Engine = require("Engine")
local Data = require("Source.Data")
local Pickup = require("Source.Pickup")

local Actor = Engine.Actor

local Item = {}

Item.ID = ""
Item.count = 1
Item.getSE = ""

function Item:init(texture, rect, tag)
    ---@cast self Source.Item
    Actor.init(self, texture, rect, tag)
    self.attributes = Data.CreateGeneralAttributeSet("Item", self.ID)
end

function Item:onCollision(other)
    ---@cast self Source.Item
    local parentCollision = super(Item, self).onCollision
    Pickup.HandleInventoryCollision(self, other, parentCollision, function (player)
        player:addItem(self.ID, self.count)
    end)
end

return class(Item, Actor)
